tools/node-modules

```diff
@@ -0,0 +1,223 @@
+}
+
+cmd_rebuild() {
+    for file in npm-shrinkwrap.json package-lock.json yarn.lock; do
+        if [ -e "${top_srcdir}/${file}" ]; then
+            echo "Please delete ${file} from ${top_srcdir} and start again"
+            exit 1
+        fi
+    done
+
+    cmd_remove
+
+    npm install
+
+    cp package.json node_modules/.package.json
+    echo /.cache >> node_modules/.gitignore
+
+    git -C node_modules init -b main
+    git -C node_modules add .
+    git -C node_modules -c gc.autoDetach=false commit -a -m "$(sha256sum package.json)"
+
+    cat <<EOF
+Next steps:
+
+  - tools/node-modules push
+
+  - git add node_modules && git commit
+
+EOF
+}
+
+cmd_push() {
+    local tag="sha-$(git -C node_modules rev-parse HEAD)"
+    git -C node_modules tag -f "${tag}" HEAD
+    git -C node_modules push "${PUSH_REMOTE}" "${tag}"
+}
+
+# called from tools/test-static-code
+cmd_test_static_code() {
+    node_modules_update=$(git log --format=%H -n 1 node_modules)
+    if ! git --no-pager log --stat --exit-code "${node_modules_update}".. package.json; then
+        cat <<EOF
+#
+# *** The above commits modify package.json without updating node_modules
+#
+# Try running:
+#
+#   tools/node-modules rebuild
+#
+EOF
+        exit 1
+    fi
+}
+
+# called from Makefile.am
+cmd_make_package_lock_json() {
+    # Run from make to ensure package-lock.json is up to date
+
+    # package-lock.json is used as the stamp file for all things that use
+    # node_modules (mostly webpack), so this is the main bit of glue that
+    # drives the entire process
+
+    # We try our best not to touch package-lock.json unless it actually changes
+
+    # This isn't going to work for a tarball, but as long as
+    # package-lock.json is already there, and newer than package.json,
+    # we're OK
+    if [ ! -e .git ]; then
+        if [ package-lock.json -nt package.json ]; then
+            exit 0
+        fi
+
+        echo "*** Can't update node modules unless running from git" >&2
+        exit 1
+    fi
+
+    # Otherwise, our main goal is to ensure that the node_modules from
+    # the index is the one that we actually have.
+    local sha="$(get_index_gitlink node_modules)"
+    if [ ! -e node_modules/.git ]; then
+        # nothing there yet...
+        cmd_checkout
+    elif [ "$(git -C node_modules rev-parse HEAD)" != "${sha}" ]; then
+        # wrong thing there...
+        cmd_checkout
+    fi
+
+    # This check is more about catching local changes to package.json than
+    # about validating something we just checked out:
+    if ! cmp -s node_modules/.package.json package.json; then
+        cat 2>&1 <<EOF
+*** package.json is out of sync with node_modules
+*** If you modified package.json, please run:
+***
+***    tools/node-modules rebuild
+***
+*** and add the result to the index.
+EOF
+        exit 1
+    fi
+
+    # Only copy the package-lock.json if it differs from the one we have
+    if ! cmp -s node_modules/.package-lock.json package-lock.json; then
```

Comment:
I don't see where `node_modules/.package-lock.json` is actually created... hmm.

Reply:
`npm` creates it.